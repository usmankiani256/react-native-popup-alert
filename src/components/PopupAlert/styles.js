import { Dimensions } from 'react-native'

const { width, height } = Dimensions.get('window')

const styles = {
  root: {
    width: width * 0.85,
    alignSelf: 'center',
  },
  animation: {
    alignSelf: 'center',
  },
  content: {
    marginHorizontal: 25,
    marginBottom: 12,
  },
  body: {
    fontSize: 17,
    lineHeight: 19,
    textAlign: 'justify',
  },
  themePrimary: {
    colors: {
      primary: '#3f3f3f',
    },
  },
  theme: {
    colors: {
      primary: '#777777',
    },
  },
}

export default styles
