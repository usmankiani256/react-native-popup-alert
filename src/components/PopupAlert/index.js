import * as React from 'react'
import { Button, Dialog, Portal } from 'react-native-paper'
import { View, Text, ScrollView, Image } from 'react-native'
import { useSelector, useDispatch } from 'react-redux'
import showAlert from '../../store/Actions/ShowAlert'
import styles from './styles'

var emitter = require('tiny-emitter/instance')

// let samplePayload = {
//   type: 'success',
//   autoDismiss: false,
//   title: 'Custom Alert Message',
//   body: 'This is a custom alert message',
//   buttons: [
//     {
//       name: 'Okay',
//       onPress: () => {},
//     },
//   ],
// }

export const showPopup = (payload) => {
  emitter.emit('Show_Alert_Fired', payload)
}

const Alert = (props) => {
  const dispatch = useDispatch()

  const SA = useSelector((state) => state.ShowAlert)
  const alert = !SA.isLoading && SA.success && SA.data

  let isSuccess = alert?.type === 'success'
  let isError = alert?.type === 'error'
  let isUnexpected = alert?.type === 'unexpected'

  React.useEffect(() => {
    emitter.on('Show_Alert_Fired', function (payload) {
      dispatch(showAlert(payload))
    })
  }, [])

  React.useEffect(() => {
    if (alert && alert.autoDismiss) {
      console.debug('Auto Dismiss after 6 seconds.')
      setTimeout(() => {
        console.debug('Auto Dismissed.')
        hideDialog()
      }, 6000)
    }
  }, [alert])

  function hideDialog() {
    dispatch(showAlert(null))
  }

  if (alert) {
    let title = alert.title || 'Custom Alert'
    let body = alert.body || 'This is a custom alert description'

    return (
      <Portal>
        <Dialog
          dismissable={false}
          style={styles.root}
          visible
          onDismiss={hideDialog}
        >
          <Dialog.Title>{title}</Dialog.Title>
          <ScrollView>
            <View style={styles.content}>
              {(isSuccess || isError || isUnexpected) && (
                <Image
                  style={{
                    ...styles.animation,
                    height: isUnexpected ? 150 : 80,
                    width: isUnexpected ? 150 : 80,
                    marginBottom: isUnexpected ? 0 : 25,
                    marginTop: isUnexpected ? -5 : 0,
                  }}
                  source={
                    isSuccess
                      ? require('../../../src/images/success.png')
                      : isError
                      ? require('../../../src/images/error.png')
                      : require('../../../src/images/unexpected.png')
                  }
                />
              )}
              <Text style={styles.body}>{body}</Text>
            </View>
            {alert.buttons && (
              <Dialog.Actions>
                {alert.buttons.map(({ name, onPress }, index) => {
                  let len = alert.buttons.length

                  return (
                    <Button
                      theme={
                        index === len - 1 ? styles.themePrimary : styles.theme
                      }
                      key={index}
                      onPress={onPress}
                    >
                      {name}
                    </Button>
                  )
                })}
              </Dialog.Actions>
            )}
          </ScrollView>
        </Dialog>
      </Portal>
    )
  } else {
    return null
  }
}

export default Alert
