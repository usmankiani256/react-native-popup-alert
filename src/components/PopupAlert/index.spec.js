import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import PopupAlert, { showPopup } from './index.js'

let mock = {
  alert: {
    icon: 'alert-circle',
    iconSize: 70,
    iconColor: 'orange',
    title: 'Custom Alert',
    body: 'This is a custom alert for testing the Alert component',
    buttons: [
      {
        name: 'Cancel',
        onPress: () => {
          showPopup(null)
        },
      },
      {
        name: 'Okay',
        onPress: () => {
          showPopup(null)
        },
      },
    ],
  },
}

describe('<PopupAlert />', () => {
  it('should match snapshot: null', () => {
    let wrapper = shallow(<PopupAlert />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot: alert', () => {
    let wrapper = shallow(<PopupAlert alert={mock.alert} />)

    expect(wrapper).toMatchSnapshot()
  })

  /* Other Cases:
    - Renders icon
    - Contains icon size
    - icon is of color passed
    - Renders title
    - Renders body
    - Renders buttons
  
  */
})
