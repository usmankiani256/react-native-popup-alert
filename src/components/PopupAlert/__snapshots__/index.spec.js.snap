// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PopupAlert /> should match snapshot: alert 1`] = `
<withTheme(Portal)
  testID="alert"
>
  <withTheme(Dialog)
    dismissable={false}
    style={
      Object {
        "alignSelf": "center",
        "width": 637.5,
      }
    }
    visible={true}
  >
    <withTheme(Dialog.Title)>
      Custom Alert
    </withTheme(Dialog.Title)>
    <ScrollView>
      <View
        style={
          Object {
            "marginBottom": 12,
            "marginHorizontal": 25,
          }
        }
      >
        <View
          style={
            Object {
              "alignSelf": "center",
            }
          }
        >
          <withTheme(IconButton)
            color="orange"
            icon="alert-circle"
            size={70}
          />
        </View>
        <Text
          style={
            Object {
              "fontSize": 17,
              "lineHeight": 19,
              "textAlign": "justify",
            }
          }
        >
          This is a custom alert for testing the Alert component
        </Text>
      </View>
      <Dialog.Actions>
        <withTheme(Button)
          key="0"
          onPress={[Function]}
          theme={
            Object {
              "colors": Object {
                "primary": "#7777",
              },
            }
          }
        >
          Cancel
        </withTheme(Button)>
        <withTheme(Button)
          key="1"
          onPress={[Function]}
          theme={
            Object {
              "colors": Object {
                "primary": "#3f3f3f",
              },
            }
          }
        >
          Okay
        </withTheme(Button)>
      </Dialog.Actions>
    </ScrollView>
  </withTheme(Dialog)>
</withTheme(Portal)>
`;

exports[`<PopupAlert /> should match snapshot: null 1`] = `""`;
